{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "# from BCBio import GFF\n",
    "from gtfparse import read_gtf\n",
    "import pprint\n",
    "import csv\n",
    "import Bio.SeqUtils\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "from Bio.SeqUtils.ProtParam import molecular_weight\n",
    "\n",
    "# from BCBio.GFF import GFFExaminer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: chr22\n",
      "Name: chr22\n",
      "Description: chr22\n",
      "Number of features: 0\n",
      "Seq('NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN')\n"
     ]
    }
   ],
   "source": [
    "# reading the fasta file (one record)\n",
    "seq=[]\n",
    "seq_file = SeqIO.read(\"data/chr22.fa\", \"fasta\")\n",
    "print(seq_file)\n",
    "seq=seq_file.seq\n",
    "\n",
    "seq_length=len(seq)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gene_type', 'gene_name', 'transcript_type', 'transcript_name', 'exon_number', 'exon_id', 'level', 'protein_id', 'transcript_support_level', 'hgnc_id', 'tag', 'ccdsid', 'havana_gene', 'havana_transcript']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "df = read_gtf(\"data/newgeneannotations.gtf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_df = df.sort_values(by=[\"gene_name\", \"start\" , \"end\"])  # Sort by gene name, and then by start and end position ascendingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_end_strand=sorted_df[[\"start\",\"end\",\"strand\",\"gene_name\",\"transcript_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 12 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n",
      "42692892\n",
      "42693950\n",
      "C:\\Users\\omars\\anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py:1676: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(ilocs[0], value, pi)\n"
     ]
    }
   ],
   "source": [
    "# De-crement indices by 1 for zero-index code\n",
    "\n",
    "start_end_strand.loc[:, \"start\"] -= 1\n",
    "start_end_strand.loc[:, \"end\"] -= 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "Dictionary = { \"Gene Seq\": [] , \"Protein Name\": [] , \"Protein Seq\" : [] , \"Mass\": [] }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "Codontable = {\n",
    "\n",
    "    \"TTT\":\"Phe\", \"TTC\":\"Phe\", \"TTA\": \"Leu\", \"TTG\":\"Leu\",\n",
    "    \"CTT\":\"Leu\", \"CTC\":\"Leu\", \"CTA\": \"Leu\", \"CTG\":\"Leu\",\n",
    "    \"ATT\":\"Ile\", \"ATC\":\"Ile\", \"ATA\": \"Ile\", \"ATG\":\"Met\",\n",
    "    \"GTT\":\"Val\", \"GTC\":\"Val\", \"GTA\": \"Val\", \"GTG\":\"Val\",\n",
    "\n",
    "\n",
    "    \"TCT\":\"Ser\", \"TCC\":\"Ser\", \"TCA\": \"Ser\", \"TCG\":\"Ser\",\n",
    "    \"CCT\":\"Pro\", \"CCC\":\"Pro\", \"CCA\": \"Pro\", \"CCG\":\"Pro\",\n",
    "    \"ACT\":\"Thr\", \"ACC\":\"Thr\", \"ACA\": \"Thr\", \"ACG\":\"Thr\",\n",
    "    \"GCT\":\"Ala\", \"GCC\":\"Ala\", \"GCA\": \"Ala\", \"GCG\":\"Ala\",\n",
    "\n",
    "\n",
    "    \"TAT\":\"Tyr\", \"TAC\":\"Tyr\", \"TAA\": \"\",\"TAG\":\"\",\n",
    "    \"CAT\":\"His\", \"CAC\":\"His\", \"CAA\": \"Gln\", \"CAG\":\"Gln\",\n",
    "    \"AAT\":\"Asn\", \"AAC\":\"Asn\", \"AAA\": \"Lys\", \"AAG\":\"Lys\",\n",
    "    \"GAT\":\"Asp\", \"GAC\":\"Asp\", \"GAA\": \"Glu\", \"GAG\":\"Glu\",\n",
    "\n",
    "\n",
    "    \"TGT\":\"Cys\", \"TGC\":\"Cys\", \"TGA\": \"\",\"TGG\":\"Trp\",\n",
    "    \"CGT\":\"Arg\", \"CGC\":\"Arg\", \"CGA\": \"Arg\", \"CGG\":\"Arg\",\n",
    "    \"AGT\":\"Ser\", \"AGC\":\"Ser\", \"AGA\": \"Arg\", \"AGG\":\"Arg\",\n",
    "    \"GGT\":\"Gly\", \"GGC\":\"Gly\", \"GGA\": \"Gly\", \"GGG\":\"Gly\",\n",
    "    \"GGN\":\"Gly\", \"NNN\": \"XXX\"\n",
    "\n",
    "\n",
    "}\n",
    "\n",
    "CodonOneCodetable = {\n",
    "\n",
    "    \"Ala\": \"A\" , \"Arg\": \"R\" , \"Gly\": \"G\" ,\n",
    "    \"Asp\": \"D\" , \"Asx\": \"B\" , \"Cys\": \"C\" ,\n",
    "    \"Gln\": \"Q\" , \"Glu\": \"E\" , \"Asn\": \"N\" ,\n",
    "    \"Glx\": \"Z\" , \"Gly\": \"G\" , \"His\": \"H\" ,\n",
    "    \"Ile\": \"I\" , \"Leu\": \"L\" , \"Lys\": \"K\" ,\n",
    "    \"Met\": \"M\" , \"Phe\": \"F\" , \"Pro\": \"P\" ,\n",
    "    \"Ser\": \"S\" , \"Thr\": \"T\" , \"Trp\": \"W\" ,\n",
    "    \"Tyr\": \"Y\" , \"Val\": \"V\" , \"XXX\": \"X\" ,\n",
    "    \"\" : \"\" \n",
    "\n",
    "\n",
    "}\n",
    "MassTable = {\n",
    "    \"A\": 89 , \"R\": 174 , \"G\": 132 ,\n",
    "    \"D\": 133, \"B\": 133 , \"C\": 121 ,\n",
    "    \"Q\": 146 ,\"E\": 147 ,\"N\": 132  ,\n",
    "    \"Z\": 147, \"G\": 147 , \"H\": 155 ,\n",
    "    \"I\": 131, \"L\": 131 , \"K\": 146 ,\n",
    "    \"M\": 149, \"F\": 165 , \"P\": 115 ,\n",
    "    \"S\": 105, \"T\": 119 , \"W\": 204 ,\n",
    "    \"Y\": 181, \"V\": 117 , \"X\": 0 \n",
    "    \n",
    "}\n",
    "\n",
    "def TranslateManual(Sequence):\n",
    "\n",
    "    protein = \"\"\n",
    "    for i in range(0, len(Sequence) , 3):\n",
    "            codon = Sequence[i:i + 3]\n",
    "            if codon[2] == \"N\" or codon[1] == \"N\" or codon[0] == \"N\":\n",
    "                protein += \"X\"\n",
    "            else:\n",
    "                protein += CodonOneCodetable[Codontable[codon.upper()]]\n",
    "    return protein\n",
    "\n",
    "def CalculateMassManual(Sequence):\n",
    "    Sum = 0\n",
    "    for i in range(0, len(Sequence)):\n",
    "            protein = Sequence[i]\n",
    "            Sum += MassTable[protein]\n",
    "    return Sum\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Protein=[]\n",
    "Gene_Region=[]\n",
    "for i in range(len(start_end_strand)):\n",
    "\n",
    "    if start_end_strand['strand'].iloc[i]==\"+\":\n",
    "        Gene_Region.append( str(seq[start_end_strand['start'].iloc[i]:start_end_strand['end'].iloc[i]]) )\n",
    "\n",
    "    elif start_end_strand['strand'].iloc[i]==\"-\":\n",
    "        Gene_Region.append( str( seq[start_end_strand['start'].iloc[i]:start_end_strand['end'].iloc[i]].reverse_complement() ) )\n",
    "\n",
    "    # Break condition\n",
    "    if i == len(start_end_strand)-1:\n",
    "        Merged_Gene_Region = ''.join(Gene_Region)  # Append all sequences of coding into 1 sequence\n",
    "\n",
    "        # Check for adding trailing N-sequence to make translation correct, is un-needed due to resulting in just an additional X at the end of the protein sequence\n",
    "        #--------------------------------\n",
    "        remainder = len(Merged_Gene_Region) % 3\n",
    "        if remainder != 0:\n",
    "            Merged_Gene_Region = Merged_Gene_Region + 'N' * (3 - remainder)\n",
    "        #------------------------------------------------------------\n",
    "        Merged_Gene_Region = Seq(Merged_Gene_Region)\n",
    "        #Protein=Merged_Gene_Region.translate(to_stop=True,cds=True)\n",
    "        Protein = TranslateManual(str(Merged_Gene_Region))\n",
    "\n",
    "\n",
    "        Dictionary[\"Protein Name\"].append(start_end_strand['gene_name'].iloc[i])\n",
    "        Dictionary[\"Protein Seq\"].append(Protein)\n",
    "        Dictionary[\"Gene Seq\"].append(Merged_Gene_Region)\n",
    "\n",
    "        #analysed_seq = ProteinAnalysis(str(Protein))\n",
    "        #Dictionary['Mass'].append(analysed_seq.molecular_weight())\n",
    "        massValue = CalculateMassManual(Protein)\n",
    "        Dictionary['Mass'].append(massValue)\n",
    "\n",
    "        Gene_Region = [ ] # Reset Gene_Region for new protein\n",
    "        Merged_Gene_Region = [ ]\n",
    "        break\n",
    "\n",
    "    # Translating Entire Gene Condition\n",
    "    if start_end_strand['gene_name'].iloc[i] != start_end_strand['gene_name'].iloc[i+1]:\n",
    "        Merged_Gene_Region = ''.join(Gene_Region)  # Append all sequences of coding into 1 sequence\n",
    "        # Check for adding trailing N-sequence to make translation correct, is un-needed due to resulting in just an additional X at the end of the protein sequence\n",
    "        #--------------------------------\n",
    "        remainder = len(Merged_Gene_Region) % 3\n",
    "        if remainder != 0:\n",
    "            Merged_Gene_Region = Merged_Gene_Region + 'N' * (3 - remainder)\n",
    "        #------------------------------------------------------------\n",
    "        Merged_Gene_Region = Seq(Merged_Gene_Region)\n",
    "        #Protein=Merged_Gene_Region.translate(to_stop=True,cds=True)\n",
    "        Protein = TranslateManual(str(Merged_Gene_Region))\n",
    "\n",
    "\n",
    "\n",
    "        Dictionary[\"Protein Name\"].append(start_end_strand['gene_name'].iloc[i])\n",
    "        Dictionary[\"Protein Seq\"].append(Protein)\n",
    "        Dictionary[\"Gene Seq\"].append(Merged_Gene_Region)\n",
    "\n",
    "        #analysed_seq = ProteinAnalysis(str(Protein))\n",
    "        #Dictionary['Mass'].append(analysed_seq.molecular_weight())\n",
    "        massValue = CalculateMassManual(Protein)\n",
    "        Dictionary['Mass'].append(massValue)\n",
    "        \n",
    "        Gene_Region = [ ] # Reset Gene_Region for new protein\n",
    "        Merged_Gene_Region = [ ]\n",
    "\n",
    "#     print(\"%0.2f\" % molecular_weight(Gene_Region,\"DNA\"))\n",
    "#     Dictionary[\"Mass\"].append(\"%0.2f\" % molecular_weight(Gene_Region,\"DNA\")) #sha8ala bs el size bydrb\n",
    "\n",
    "\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "Data=pd.DataFrame.from_dict(Dictionary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Seq</th>\n",
       "      <th>Protein Name</th>\n",
       "      <th>Protein Seq</th>\n",
       "      <th>Mass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>(G, G, C, A, G, T, A, A, T, G, A, A, A, C, T, ...</td>\n",
       "      <td>CELSR1</td>\n",
       "      <td>GSNETSIHLEKSHLPAAADADGADAEGPAPGEAGRLAEPHILAHVF...</td>\n",
       "      <td>394401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>(A, T, G, G, C, C, A, T, C, T, C, A, T, C, A, ...</td>\n",
       "      <td>MYO18B</td>\n",
       "      <td>MAISSRLALWEQKFGKRTRALHHPRPLLFSLSSQGASLSNWSGGLK...</td>\n",
       "      <td>334956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373</th>\n",
       "      <td>(T, T, C, T, A, C, T, G, A, G, C, T, A, C, C, ...</td>\n",
       "      <td>TTC28</td>\n",
       "      <td>FYATQAPQPQLLLPRVSCFCSVCFQRPALGLPATPLLSHRCGQHRL...</td>\n",
       "      <td>319153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>(A, T, G, G, C, C, G, A, G, A, A, T, G, T, G, ...</td>\n",
       "      <td>EP300</td>\n",
       "      <td>MAENVVEPGPPSAKRPKLSSPALSASASDGTILALYLTWSTTYQMN...</td>\n",
       "      <td>311798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>(A, T, G, G, A, G, G, A, G, G, T, G, C, C, T, ...</td>\n",
       "      <td>TRIOBP</td>\n",
       "      <td>MEEVPGDALCEHFEANILTQNRCQNCFHPEEAHGARYQSSGALQVL...</td>\n",
       "      <td>304768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>(A, T, G, G, C, G, T, C, C, G, G, A, G, C, G, ...</td>\n",
       "      <td>SMDT1</td>\n",
       "      <td>MASGAARWLVLAPVRSGALRSGPSLRKDGDVSAAWSGSGRSLVPSR...</td>\n",
       "      <td>13694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>(G, C, G, C, C, G, C, C, A, A, G, A, T, G, A, ...</td>\n",
       "      <td>DNAL4</td>\n",
       "      <td>APPRSKRQWTRSSAPPGTWSARALGLRSPTRRTSSTCTSGAPWLCA...</td>\n",
       "      <td>13346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>(T, G, G, C, T, C, C, A, T, T, T, G, T, C, C, ...</td>\n",
       "      <td>ATP5MGL</td>\n",
       "      <td>WLHLSVTLWRRPQHWMLLLTSLDWPHFGTTPRLSWFLPPLLRSLEL...</td>\n",
       "      <td>12828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>(G, C, A, G, G, C, G, A, C, C, T, G, G, A, T, ...</td>\n",
       "      <td>SYCE3</td>\n",
       "      <td>AGDLDGLHGGDAHQPYAGRVHASAGGCLRQLQGGDGEELARAAADQ...</td>\n",
       "      <td>11329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>(A, T, G, G, C, G, G, C, C, G, C, G, A, C, G, ...</td>\n",
       "      <td>UQCR10</td>\n",
       "      <td>MAAATLTSKLYSLLFRRTSTFALTIIVGVMFFERAFDQGADAIYDH...</td>\n",
       "      <td>8166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              Gene Seq Protein Name  \\\n",
       "64   (G, G, C, A, G, T, A, A, T, G, A, A, A, C, T, ...       CELSR1   \n",
       "207  (A, T, G, G, C, C, A, T, C, T, C, A, T, C, A, ...       MYO18B   \n",
       "373  (T, T, C, T, A, C, T, G, A, G, C, T, A, C, C, ...        TTC28   \n",
       "111  (A, T, G, G, C, C, G, A, G, A, A, T, G, T, G, ...        EP300   \n",
       "367  (A, T, G, G, A, G, G, A, G, G, T, G, C, C, T, ...       TRIOBP   \n",
       "..                                                 ...          ...   \n",
       "322  (A, T, G, G, C, G, T, C, C, G, G, A, G, C, G, ...        SMDT1   \n",
       "100  (G, C, G, C, C, G, C, C, A, A, G, A, T, G, A, ...        DNAL4   \n",
       "31   (T, G, G, C, T, C, C, A, T, T, T, G, T, C, C, ...      ATP5MGL   \n",
       "336  (G, C, A, G, G, C, G, A, C, C, T, G, G, A, T, ...        SYCE3   \n",
       "386  (A, T, G, G, C, G, G, C, C, G, C, G, A, C, G, ...       UQCR10   \n",
       "\n",
       "                                           Protein Seq    Mass  \n",
       "64   GSNETSIHLEKSHLPAAADADGADAEGPAPGEAGRLAEPHILAHVF...  394401  \n",
       "207  MAISSRLALWEQKFGKRTRALHHPRPLLFSLSSQGASLSNWSGGLK...  334956  \n",
       "373  FYATQAPQPQLLLPRVSCFCSVCFQRPALGLPATPLLSHRCGQHRL...  319153  \n",
       "111  MAENVVEPGPPSAKRPKLSSPALSASASDGTILALYLTWSTTYQMN...  311798  \n",
       "367  MEEVPGDALCEHFEANILTQNRCQNCFHPEEAHGARYQSSGALQVL...  304768  \n",
       "..                                                 ...     ...  \n",
       "322  MASGAARWLVLAPVRSGALRSGPSLRKDGDVSAAWSGSGRSLVPSR...   13694  \n",
       "100  APPRSKRQWTRSSAPPGTWSARALGLRSPTRRTSSTCTSGAPWLCA...   13346  \n",
       "31   WLHLSVTLWRRPQHWMLLLTSLDWPHFGTTPRLSWFLPPLLRSLEL...   12828  \n",
       "336  AGDLDGLHGGDAHQPYAGRVHASAGGCLRQLQGGDGEELARAAADQ...   11329  \n",
       "386  MAAATLTSKLYSLLFRRTSTFALTIIVGVMFFERAFDQGADAIYDH...    8166  \n",
       "\n",
       "[408 rows x 4 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_Data = Data.sort_values(by=[\"Mass\"] , ascending=False)\n",
    "sorted_Data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_Data.to_csv(\"Output_P1_Masses.csv\" , index = False)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5a862fa755b339cb269ec5ceb6e73837f63b6953002ae94843c7ff6941dd7636"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
